### PR TITLE
feat(frontend):  web notification improvement

### PR DESCRIPTION
**Pull request by Millefeuille42**
#6428 

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [x] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

### CODE REVIEW

**Comment by Millefeuille42** [comment]
Notifications are still not sorted in sidebar space list.
TODOs and FIXMEs added in code, they need to be sorted out.

**Comment by Millefeuille42** [comment]
Web notifications are almost functional; however, the management of workspaces in the sidebar still uses an unfiltered list, and there are a few FIXMEs and TODOs that need to be addressed or decided as sufficient for now.

In detail, the complicated part is handling notifications outside the notification wall. Three components use the `notificationPage` reducer:
- The notification wall
- The `Tracim` component, which defines the badge next to the "Notifications" sidebar element, as well as the favicon and page title
- The `SidebarSpaceItem` component, which corresponds to each entry in the workspace list in the sidebar

- In the `Tracim` component, I need to filter the list in two places
- In the notification wall component, I filter once
- If we leave things as they are, it will require filtering once per space in the sidebar (I think this is problematic)

I've left "optimization" suggestions in case they're needed for version 1, given the mentioned issues.
Additionally, the issue raised in the issue still needs to be fixed, unless we can ignore it for version 1 and create a dedicated ticket for it.

File: frontend/src/container/Tracim.jsx [hunk 1]
```diff
@@ -505,14 +506,32 @@ export class Tracim extends React.Component {
...
+    })
+
+    return { unreadNotificationCount, unreadMentionCount, notificationList }
+  }
+
+  handleHeadTitleAndFavicon = (prevHeadTitleArgs, prevUnreadNotificationCount, prevUnreadMentionCount, prevUserConfig) => {
+    const { props } = this
+
+    // TODO - M.L - 2024-06-07 - Maybe use results as a state so that it is not necessary to sort twice
+    const { unreadNotificationCount, unreadMentionCount } = this.filterAndCountNotifications()
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/Tracim.jsx:530]: or the filtered list

[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/Tracim.jsx:530]: and

File: frontend_lib/src/helper.js [hunk 2]
```diff
@@ -18,6 +18,7 @@ import {
   getReservedUsernames,
   getUsernameAvailability
 } from './action.async.js'
+import { MENTION_CONSTANT } from './mentionOrLink'
```
Comment (Skylsmoi) [inline frontend_lib/src/helper.js:21]: missing .jsx

File: frontend/src/container/Tracim.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/Tracim.jsx:530]: or in a reducer

[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/Tracim.jsx:530]: check complexity to put in a reducer, it is probably the best solution unless it's very complex

File: frontend_lib/src/helper.js [hunk 3]
```diff
@@ -994,3 +995,22 @@ export const getRevisionTypeLabel = (revisionType, t) => {
...
   return revisionType
 }
+
+export const buildUserConfigContentNotifyAllKey = contentId => `content.${contentId}.notify_all_members_message`
+export const buildUserConfigSpaceWebNotificationKey = spaceId => `space.${spaceId}.web_notification`
+export const buildUserConfigContentWebNotificationKey = contentId => `content.${contentId}.web_notification`
+
+// FIXME - M.L - 2024-06-07 - Use of plain string 'mention.created' since importing TLM related constants
+//  breaks the imports (maybe circular imports ?)
+//  TODO Find a solution or explain why we kept it this way
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend_lib/src/helper.js]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend_lib/src/helper.js (hunk 2 above)
Comment (Millefeuille42) [reply frontend_lib/src/helper.js]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/container/Tracim.jsx [hunk 4]
```diff
@@ -505,14 +506,32 @@ export class Tracim extends React.Component {
...
   }
 
-  handleHeadTitleAndFavicon = (prevHeadTitleArgs, prevUnreadNotificationCount, prevUnreadMentionCount) => {
+  filterAndCountNotifications = () => {
     const { props } = this
 
+    let unreadNotificationCount = props.notificationPage.unreadNotificationCount
+    let unreadMentionCount = props.notificationPage.unreadMentionCount
+    const notificationList = props.notificationPage.list.filter(notification => {
+      const keep = shouldKeepNotification(notification, props.user.config)
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/container/Tracim.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: backend/tracim_backend/lib/core/event.py [hunk 5]
```diff
@@ -1161,6 +1197,21 @@ def _get_user_call_event_receiver_ids(
...
+    event: Event, session: TracimSession, config: CFG
+) -> Set[int]:
+    """
+    User configs are readable by
+    - subscription author
+    - administrators
+    - workspace manager of the workspace the subscription took place
+    """
+    user_api = UserApi(current_user=None, session=session, config=config)
+    administrators = user_api.get_user_ids_from_profile(Profile.ADMIN)
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply backend/tracim_backend/lib/core/event.py]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

Removed administrators

File: frontend/src/component/Account/UserSpacesConfig.jsx [hunk 6]
```diff
@@ -22,13 +22,16 @@ import {
...
+  putUserConfiguration,
+  serialize,
+  buildUserConfigSpaceWebNotificationKey
 } from 'tracim_frontend_lib'
 import { serializeUserConfig, serializeUserWorkspaceConfig, serializeWorkspaceListProps } from '../../reducer/workspaceList.js'
 import { newFlashMessage } from '../../action-creator.sync.js'
 import { deleteUserRole, getUserWorkspaceConfigList } from '../../action-creator.async.js'
 import AdminUserSpacesConfig from '../../container/AdminUserSpacesConfig.jsx'
 import UserSpacesConfigLine from './UserSpacesConfigLine.jsx'
+import { FETCH_CONFIG } from '../../util/helper'
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/Account/UserSpacesConfig.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/Account/UserSpacesConfig.jsx [hunk 7]
```diff
@@ -87,6 +90,8 @@ export const UserSpacesConfig = (props) => {
           onChangeEmailNotificationType={
             emailNotificationType => props.onChangeEmailNotificationType(space.id, emailNotificationType)
           }
+          WebNotificationEnabled={!props.admin && props.user.config[buildUserConfigSpaceWebNotificationKey(space.id)]}
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/Account/UserSpacesConfig.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/Account/UserSpacesConfig.jsx [hunk 8]
```diff
@@ -115,6 +120,25 @@ export const UserSpacesConfig = (props) => {
     })
   }
 
+  const handleChangeWebNotification = async (spaceId) => {
+    const userConfig = { ...props.user.config }
+    const currentStatus = userConfig[buildUserConfigSpaceWebNotificationKey(spaceId)]
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/Account/UserSpacesConfig.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/Account/UserSpacesConfigLine.jsx [hunk 9]
```diff
@@ -62,4 +77,19 @@ class UserSpacesConfigLine extends React.Component {
...
+  onlyManager: PropTypes.bool,
+  admin: PropTypes.bool,
+  WebNotificationEnabled: PropTypes.bool,
+  onLeaveSpace: PropTypes.func,
+  onChangeEmailNotificationType: PropTypes.func,
+  onChangeWebNotification: PropTypes.func.isRequired
+}
+
+UserSpacesConfigLine.defaultProps = {
+  WebNotificationEnabled: true
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/Account/UserSpacesConfigLine.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/Account/UserSpacesConfig.jsx [hunk 10]
```diff
@@ -264,6 +288,7 @@ export const UserSpacesConfig = (props) => {
                         />
                       </th>
                       {props.system.config.email_notification_activated && <th>{props.t('Email notifications')}</th>}
+                      <th>{props.t('Web notifications')}</th>
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/Account/UserSpacesConfig.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/WebNotificationSpaceSelector/WebNotificationSpaceSelector.jsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
+import React from 'react'
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/WebNotificationSpaceSelector/WebNotificationSpaceSelector.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/WebNotificationSpaceSelector/WebNotificationSpaceSelector.jsx [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        placement='left'
+        trigger='click'
+        popoverBody={
+          <div className='PopoverHelperBody'>
+            <div className='PopoverHelperBody__section'>
+              <div className='PopoverHelperBody__section__label'>
+                {props.t('Web notifications')}
+              </div>
+              <div className='PopoverHelperBody__section__description'>
+                <Trans>
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/WebNotificationSpaceSelector/WebNotificationSpaceSelector.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/component/WebNotificationSpaceSelector/WebNotificationSpaceSelector.jsx [hunk 13]
```diff
@@ -0,0 +1,77 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import {
+  Popover,
+  BtnSwitch
+} from 'tracim_frontend_lib'
+import { translate, Trans } from 'react-i18next'
+
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/component/WebNotificationSpaceSelector/WebNotificationSpaceSelector.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/container/Dashboard.jsx [hunk 14]
```diff
@@ -415,6 +417,27 @@ export class Dashboard extends React.Component {
     }
   }
 
+  handleClickChangeWebNotification = async () => {
+    const { props } = this
+
+    const userConfig = { ...props.user.config }
+    const currentStatus = userConfig[buildUserConfigSpaceWebNotificationKey(props.currentWorkspace.id)]
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/container/Dashboard.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

File: frontend/src/container/NotificationWall.jsx [hunk 15]
```diff
@@ -277,7 +279,7 @@ export const NotificationWall = props => {
 
     setIsFolderPathLoading(false)
     setNotificationList(newNotificationList)
-  }, [props.notificationPage.list])
+  }, [props.notificationPage.list, props.user])
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply frontend/src/container/NotificationWall.jsx]: Fixed in [9d13cc1](https://github.com/tracim/tracim/pull/6518/commits/9d13cc12cbc7970243cb0b47d750df5853fde60a)

**Review by Millefeuille42 — COMMENTED** [review]
Missing translation keys

File: backend/tracim_backend/migration/versions/38d2725ca92f_added_user_config_entity_type.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
+"""Added user_config entity type
```
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply backend/tracim_backend/migration/versions/38d2725ca92f_added_user_config_entity_type.py:1]: fixed in [2b2f9d5](https://github.com/tracim/tracim/pull/6518/commits/2b2f9d5d445372f209187d7fbe4a366c2c1f05ba)

**Comment by floreal15** [comment]
after looking at the failed test, there seem to be two bugs, one with the message to set them as read ( might be because it's done differently due to the change and need to be updated in the test) 
and another bug with deleting user, might come from the update of the user_config in the backend

File: frontend/src/container/Tracim.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/Tracim.jsx]: fixed: intelligence moved into the reducer notificationPage

File: frontend/src/container/Tracim.jsx [hunk 17]
```diff
@@ -505,14 +506,32 @@ export class Tracim extends React.Component {
     }
   }
 
-  handleHeadTitleAndFavicon = (prevHeadTitleArgs, prevUnreadNotificationCount, prevUnreadMentionCount) => {
+  filterAndCountNotifications = () => {
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/Tracim.jsx]: fixed

**Review by Skylsmoi — APPROVED** [review]

**Comment by Skylsmoi** [comment]
tested on branch

**Pull request merged**